### ISSUE TITLE
Updating an image does not work in a single environment

### ISSUE THREAD

**Issue by sapetnioc** [brainvisa/casa-distro#300]
If one creates a dev environment by downloading an image and creating the dev environment with this image (without using `~/casa_distro` directory), the `casa_distro pull_image` is not working. It cannot find the appropriate image on the site.

**Comment by sapetnioc** [comment]
In the branch `pull_image` the remote image is now found. But the update overwrite the existing image whose name contains the build number. In my case  `casa-dev-5.0-5.sif` is replaced by `casa-dev-5.0-8.sif` but without changing the file name (that would require to change the environment configuration). I do not really know what is the best option to solve this.

**Comment by denisri** [comment]
In a perfect world, given that `casa-dev-5.0-5.sif` is erased, every environment which uses it should be modified to use the new image, or better, the "more generic" `casa-dev-5.0.sif` (which should be a symlink to the other).

**Comment by sapetnioc** [comment]
In my case, `casa-dev-5.0-5.sif` should be removed and replaced by
`casa-dev-5.0-8.sif` and the configuration file updated accordingly. But
the right things to do depends on the user configuration (image shared or
not, symlink or not, etc.). I think that there are two standard use cases
that must be dealed with without need for the user to add parameters :

   - Several environments in `base_directory` (i.e. `~/casa_distro)
   - A single environment created without a `base_directory`

But these two can co-exist and this is difficult to know if an image file
is shared by several environments. We may be able to solve this problem by
changing the way we find environments. We could record all environment
paths in a `~/.config/casa_distro` file. The setup command would upate this
file (adding a new path). Hence, `casa_distro` would have a complete view
of all images and environments used and how images are shared. This way we
could provide good default behavior for `update_image` in all cases. If we
go this way, we may completely get rid of `base_directory` in favor of  an
environments config file.

**Comment by denisri** [comment]
I don't totally adhere to the `~/.config/casa_distro` file although it can be discussed (I just find it's not necessary given the other ways of finding environments, and I sometimes use several independent dirs for tests). Up to now we have the `$CASA_BASE_DIRECTORY` variable and the default fallback `~/casa_distro`. To begin with, we can follow this existing rule. In this environments directory, at least, we can list and modify environments which use images which don't exist anymore because they are replaced with newer ones.

**Comment by sapetnioc** [comment]
Well, let's make it simple and forget about the config file for this point.
The variable and default fallback was not enough for independent directories. This is what I changed in the commit above. If the default directory is not existing, the base directory is the environment directory (commands already support that the base directory can be a single environment directory). I do not really like this solution because the behavior is different if `~/casa_distro` exists or not.

What do you think of the following rule for finding the base directory:
- The default base directory is either `$CASA_BASE_DIRECTORY` or `~/casa_distro`
- If the environment is in the default directory => base directory = default base directory
   otherwise => base directory = environment directory (two parents above `bv` command).

**Comment by sapetnioc** [comment]
Given this, `casa_distro` will always see a list of environments (with only one for independent ones). 
Now, I think `pull_image` will have to be a little smarter to decide if it must overwrite the image or not. If image name contains a build number, the newer image will be downloaded next to it and environments using the old image will have their config modified to use the new one. But if there is no build number in the image name, the file will be replaced and no config will be modified.
Last, we must decide if we remove the old image. I would be in favor of keeping it and to rely on `clean_images` to remove unused ones.

**Comment by sapetnioc** [comment]
To fix this issue, I made a change to `casa_distro list` to always consider the current environment of `casa_distro` command, then consider the environments found in `~/casa_distro` (as before). Then I rewrote `casa_distro pull_image` to make its behaviour simpler to understand.

Now `pull_image` will first select a set of image files to check for update. By default, it includes all images of environments returned by the `list` command. It is possible to filter these environments (by `distro`, `branch`, `system`, etc.). It is also possible to consider a single image file with `image=<file>`.

Once images are selected, casa-distro environments returned by `list` are parsed to identify config files (i.e. `casa_distro.json`) to modify after image updates.

Finally, for each considered image file, an updated version is searched on the web site and, if found, it is downloaded . The corresponding config files are modified. The original image is deleted (can be kept with `cleanup=no`).

It is possible to do a dry run and simply check what the command would do by using `mode=fake`.
